File: jssdk/src/services/JobsService.ts
import { BaseService } from "@/services/BaseService";
import { CommonOptions } from "@/tools/options";

/**
 * Job 状态类型
 */
export type JobStatus = "pending" | "processing" | "completed" | "failed";

/**
 * Job 模型
 */
export interface Job {
    id: string;
    topic: string;
    payload: Record<string, any>;
    status: JobStatus;
    run_at: string;
    locked_until?: string;
    retries: number;
    max_retries: number;
    last_error?: string;
    created: string;
    updated: string;
}

/**
 * 入队选项
 */
export interface JobEnqueueOptions {
    /** 调度时间（延时执行），ISO 8601 格式 */
    run_at?: string | Date;
    /** 最大重试次数 */
    max_retries?: number;
}

/**
 * 列表查询选项
 */
export interface JobListOptions extends CommonOptions {
    /** 按 topic 筛选 */
    topic?: string;
    /** 按状态筛选 */
    status?: JobStatus;
    /** 返回数量限制 */
    limit?: number;
    /** 偏移量 */
    offset?: number;
}

/**
 * 列表查询结果
 */
export interface JobListResult {
    items: Job[];
    total: number;
    limit: number;
    offset: number;
}

/**
 * 统计信息
 */
export interface JobStats {
    pending: number;
    processing: number;
    completed: number;
    failed: number;
    total: number;
    success_rate: number;
}

/**
 * JobsService 提供 Jobs API 的客户端封装
 */
export class JobsService extends BaseService {
    /**
     * 入队一个新任务
     *
     * @param topic 任务主题
     * @param payload 任务数据
     * @param options 入队选项
     * @throws {ClientResponseError}
     */
    async enqueue(
        topic: string,
        payload: Record<string, any>,
        options?: JobEnqueueOptions & CommonOptions,
    ): Promise<Job> {
        const body: Record<string, any> = {
            topic,
            payload,
        };

        if (options?.run_at) {
            body.run_at =
                options.run_at instanceof Date
                    ? options.run_at.toISOString()
                    : options.run_at;
        }

        if (options?.max_retries !== undefined) {
            body.max_retries = options.max_retries;
        }

        const sendOptions = Object.assign(
            {
                method: "POST",
                body,
            },
            options,
        );

        // 移除非 SendOptions 的属性
        delete (sendOptions as any).run_at;
        delete (sendOptions as any).max_retries;

        return this.client.send("/api/jobs/enqueue", sendOptions);
    }

    /**
     * 获取单个任务详情
     *
     * @param id 任务 ID
     * @param options 请求选项
     * @throws {ClientResponseError}
     */
    async get(id: string, options?: CommonOptions): Promise<Job> {
        const sendOptions = Object.assign(
            {
                method: "GET",
            },
            options,
        );

        return this.client.send(`/api/jobs/${encodeURIComponent(id)}`, sendOptions);
    }

    /**
     * 获取任务列表
     *
     * @param options 列表查询选项
     * @throws {ClientResponseError}
     */
    async list(options?: JobListOptions): Promise<JobListResult> {
        const query: Record<string, any> = {};

        if (options?.topic) {
            query.topic = options.topic;
        }
        if (options?.status) {
            query.status = options.status;
        }
        if (options?.limit !== undefined) {
            query.limit = options.limit;
        }
        if (options?.offset !== undefined) {
            query.offset = options.offset;
        }

        const sendOptions = Object.assign(
            {
                method: "GET",
                query,
            },
            options,
        );

        // 移除非 SendOptions 的属性
        delete (sendOptions as any).topic;
        delete (sendOptions as any).status;
        delete (sendOptions as any).limit;
        delete (sendOptions as any).offset;

        return this.client.send("/api/jobs", sendOptions);
    }

    /**
     * 重新入队一个失败的任务
     *
     * @param id 任务 ID
     * @param options 请求选项
     * @throws {ClientResponseError}
     */
    async requeue(id: string, options?: CommonOptions): Promise<Job> {
        const sendOptions = Object.assign(
            {
                method: "POST",
            },
            options,
        );

        return this.client.send(
            `/api/jobs/${encodeURIComponent(id)}/requeue`,
            sendOptions,
        );
    }

    /**
     * 删除一个任务（仅 pending 或 failed 状态可删除）
     *
     * @param id 任务 ID
     * @param options 请求选项
     * @throws {ClientResponseError}
     */
    async delete(id: string, options?: CommonOptions): Promise<boolean> {
        const sendOptions = Object.assign(
            {
                method: "DELETE",
            },
            options,
        );

        return this.client
            .send(`/api/jobs/${encodeURIComponent(id)}`, sendOptions)
            .then(() => true);
    }

    /**
     * 获取任务统计信息
     *
     * @param options 请求选项
     * @throws {ClientResponseError}
     */
    async stats(options?: CommonOptions): Promise<JobStats> {
        const sendOptions = Object.assign(
            {
                method: "GET",
            },
            options,
        );

        return this.client.send("/api/jobs/stats", sendOptions);
    }
}

File: pythonsdk/src/pocketbase/services/jobs_service.py
"""JobsService - Service for managing PocketBase background jobs."""

from typing import TYPE_CHECKING, Any
from urllib.parse import quote, urlencode

from pocketbase.services.base_service import BaseService

if TYPE_CHECKING:
    from pocketbase.client import PocketBase


class JobsService(BaseService):
    """Service for managing PocketBase background jobs.
    
    Provides methods to list, get, enqueue, requeue, delete jobs and get stats.
    """

    def __init__(self, client: "PocketBase") -> None:
        """Initialize the JobsService.
        
        Args:
            client: The PocketBase client instance.
        """
        super().__init__(client)

    def get_list(
        self,
        limit: int = 30,
        offset: int = 0,
        topic: str = "",
        status: str = ""
    ) -> dict[str, Any]:
        """Get list of jobs with optional filtering.
        
        Args:
            limit: Maximum number of jobs to return.
            offset: Number of jobs to skip.
            topic: Filter by topic name.
            status: Filter by job status (pending, running, completed, failed).
            
        Returns:
            Dict with items and total count.
            
        Raises:
            ClientResponseError: If the request fails.
        """
        params: dict[str, Any] = {
            "limit": limit,
            "offset": offset,
        }
        if topic:
            params["topic"] = topic
        if status:
            params["status"] = status
        
        return self._client.send("/api/jobs", method="GET", params=params)

    def get_one(self, job_id: str) -> dict[str, Any]:
        """Get a single job by ID.
        
        Args:
            job_id: The job ID.
            
        Returns:
            The job data.
            
        Raises:
            ClientResponseError: If the request fails or job not found.
        """
        return self._client.send(f"/api/jobs/{quote(job_id, safe='')}", method="GET")

    def enqueue(
        self,
        topic: str,
        payload: dict[str, Any] | None = None,
        run_at: str | None = None,
        max_retries: int | None = None
    ) -> dict[str, Any]:
        """Enqueue a new job.
        
        Args:
            topic: The job topic.
            payload: The job payload data.
            run_at: Optional ISO datetime string for delayed execution.
            max_retries: Maximum number of retries on failure.
            
        Returns:
            The created job data.
            
        Raises:
            ClientResponseError: If the request fails.
        """
        body: dict[str, Any] = {
            "topic": topic,
            "payload": payload or {},
        }
        if run_at:
            body["run_at"] = run_at
        if max_retries is not None:
            body["max_retries"] = max_retries
        
        return self._client.send("/api/jobs/enqueue", method="POST", body=body)

    def requeue(self, job_id: str) -> dict[str, Any]:
        """Requeue a failed job.
        
        Args:
            job_id: The job ID to requeue.
            
        Returns:
            The requeued job data.
            
        Raises:
            ClientResponseError: If the request fails.
        """
        return self._client.send(
            f"/api/jobs/{quote(job_id, safe='')}/requeue",
            method="POST"
        )

    def delete(self, job_id: str) -> bool:
        """Delete a job.
        
        Args:
            job_id: The job ID to delete.
            
        Returns:
            True if deletion was successful.
            
        Raises:
            ClientResponseError: If the request fails.
        """
        self._client.send(f"/api/jobs/{quote(job_id, safe='')}", method="DELETE")
        return True

    def get_stats(self) -> dict[str, Any]:
        """Get job statistics.
        
        Returns:
            Dict with job counts by status.
            
        Raises:
            ClientResponseError: If the request fails.
        """
        return self._client.send("/api/jobs/stats", method="GET")
